### PR TITLE
Add auditing to AutomaticUpdateRule model

### PR DESCRIPTION
**Pull request by gherceg**
## Technical Summary

https://dimagi-dev.atlassian.net/browse/SAAS-13015

Adds auditing of `active` and `deleted` fields of the `AutomaticUpdateRule` model object. This object is used in conditional alerts, case update rules, and case deduplication rules.

This does not include the bootstrapping of existing auto update rules because we need to run that as a separate management command post-deploy after the audit decorator is setup on `AutomaticUpdateRule`. I created https://github.com/dimagi/commcare-cloud/pull/5611 on commcare-cloud which creates changelog entry for running this management command as well. 

The management command that will be run post deploy is `manage.py bootstrap_field_audit_events top-up AutomaticUpdateRule`

### Safety story

Auditing is not user facing. This is a well-tested library, and includes unit tests to ensure `AutomaticUpdateRule` changes are audited as expected.

### Automated test coverage

`corehq/apps/data_interfaces/tests/test_auto_update_rules.py` as well as tests in [django-field-audit](https://github.com/dimagi/django-field-audit).
### QA Plan

No QA necessary.

### Rollback instructions

Reverting this PR will remove the creation of audit events for the AutomaticUpdateRule model, and would require re-running the bootstrap command if this work is re-deployed.

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/data_interfaces/models.py [hunk 1]
```diff
@@ -84,6 +85,7 @@ def _try_date_conversion(date_or_string):
     return date_or_string
 
 
+@audit_fields("active", "deleted")
```
Comment (joxl) [inline corehq/apps/data_interfaces/models.py:88]: Not sure if you've looked into the `audit_special_queryset_writes` keyword argument here, but we generally always want it in order to ensure audit events can never be accidentally missed if someone uses any of `.bulk_create()`, `.bulk_update()`, `.delete()` or `.update()` QuerySet methods.  Adding it comes with the extra work of ensuring that the `audit_action` keyword argument gets specified anywhere these methods are currently used in HQ (for this model).

```suggestion
@audit_fields("active", "deleted", audit_special_queryset_writes=True)
```

**Review by joxl — COMMENTED** [review]
One comment regarding special QuerySet methods.

For other reviewers: I offlined with @gherceg and I'm going to work on a "django-field-audit HOWTO doc" for HQ to help clarify some of the recommended usage patterns in HQ for devs who aren't already deeply familiar with this utility.

**Comment by gherceg** [comment]
Thanks for the feedback @joxl. Joel and I also discussed potentially auditing additional fields (like domain, case_type, etc) to enable ease of querying. `domain` and `case_type` _should_ always be constant in this case, but it is a matter of pulling this data solely from the `AuditEvent` table vs a JOIN with the `AutomaticUpdateRule` table. Perhaps more importantly, if an `AutomaticUpdateRule` is permanently deleted and we have any intention of still making use of the past audit events associated with that object, we would **have to** audit fields like `domain` and `case_type`. Going to ask this on the ticket to confirm which approach we should take before proceeding.

File: corehq/apps/data_interfaces/models.py (hunk 1 above)
Comment (gherceg) [reply corehq/apps/data_interfaces/models.py]: See cdbb3cbd8d87eb1cd9c81515fdbc69bbf42cb91c (though I did something bad to tests 😅)

**Review by joxl — APPROVED** [review]
Looks good!

**Comment by gherceg** [comment]
I realized there are use cases of `AutomaticUpdateRule.objects.filter(...).update()` which is not yet supported by django-field-audit. My initial reaction was to change those `QuerySet.update` calls to `Model.update`, especially since they appeared to be [specific to one instance](https://github.com/dimagi/commcare-hq/blob/7073f2f0d6fe4f70c50960ff6eb0d06285bc696c/corehq/apps/data_interfaces/deduplication.py#L174-L177). However it probably makes more sense to just update django-field-audit to support auditing QuerySet.update. Looking into that.

**Comment by gherceg** [comment]
Created https://github.com/dimagi/django-field-audit/pull/11 to support `QuerySet.update()` in django-field-audit

File: corehq/apps/data_interfaces/deduplication.py [hunk 2]
```diff
@@ -171,10 +171,8 @@ def backfill_deduplicate_rule(domain, rule):
         )
     finally:
         progress_helper.set_rule_complete()
-        AutomaticUpdateRule.objects.filter(pk=rule.pk).update(
-            locked_for_editing=False,
```
Comment (snopoke) [inline corehq/apps/data_interfaces/deduplication.py:175]: doesn't this also need to be updated?

Comment (gherceg) [reply corehq/apps/data_interfaces/deduplication.py:175]: Thanks Simon. I only checked for `AutomaticUpdateRule.by_model` 🤦🏻. 

Also curious to hear if you have strong opinions on `AutomaticUpdateRule.objects.filter(rule=rule.pk).update(...)` vs `rule.save(update_field=[...])`. I figured in the cases where we already had access to the `rule` db object, using the latter method made sense, but open to changes.

**Comment by gherceg** [comment]
Curious to hear if anyone has strong opinions on `AutomaticUpdateRule.objects.filter(rule=rule.pk).update(...)` vs `rule.save(update_fields=[...])`. I figured in the cases where we already had access to the rule db object, using the latter method made sense, but open to changes.

**Comment by joxl** [comment]
> opinions on `AutomaticUpdateRule.objects.filter(rule=rule.pk).update(...)` vs `rule.save(update_fields=[...])`.

In general, `QuerySet.update()` can prevent an extra fetch from the db. If the model is already fetched then the difference probably isn't important, unless you're concerned about signals.  The `save(update_fields=[...])` will fire `pre_save()` signals (which is fewer signals than `save()` typically fires), while `QuerySet.update(...)` will never fire any signals.

There are also nuances depending on which fields are being updated.  The `save()` method will perform "auto value logic" to populate a fields with an automatic value if needed (for example, date/time columns with `auto_now=True`).  The `QuerySet.update()` method will not do this for you automatically. Another clue about possible intended uses for `save(update_fields=[...])` is that if `update_fields` is an empty iterable then the save will be skipped entirely, possibly making it a clever "update maybe" tool.

File: corehq/apps/data_interfaces/deduplication.py (hunk 2 above)
Comment (snopoke) [reply corehq/apps/data_interfaces/deduplication.py:175]: This still isn't setting `locked_for_editing` to False like it was before

**Review by snopoke — CHANGES_REQUESTED** [review]

**Review by joxl — APPROVED** [review]

**Comment by gherceg** [comment]
Closed and re-opned to try to get this task list to complete, but not looking promising 😞 the description previously had an unchecked box, but I have since deleted it.

File: corehq/apps/data_interfaces/deduplication.py (hunk 2 above)
Comment (gherceg) [reply corehq/apps/data_interfaces/deduplication.py:175]: See 1e2cc960a39e3513473e39d055929486f779a320

**Review by snopoke — APPROVED** [review]
Approved but it looks like there still an outstanding question about auditing `domain` and `case_type`: https://github.com/dimagi/commcare-hq/pull/32336#issuecomment-1312151039

**Comment by gherceg** [comment]
> Approved but it looks like there still an outstanding question about auditing domain and case_type: https://github.com/dimagi/commcare-hq/pull/32336#issuecomment-1312151039

Good catch. This never got answered in the ticket and I completely forgot about it! After talking over with Cal and Joel, there is no reason _not_ to add fields that don't change often or at all (think domain and workflow). I'll add one more commit to add those fields to the auditing decorator.

**Review by joxl — APPROVED** [review]

**Review by snopoke — APPROVED** [review]

**Comment by gherceg** [comment]
Slight change of plans. Rather than run the management command manually and require a new changelog entry in commcare-cloud, I will create a PR that runs the management command as part of a migration, and merge that PR 3 weeks after this PR, which will allow time for self hosters to deploy this change prior to the migration.

The concern around including the migration in the same deploy as the initial auditing of changes is that the migration will be run on the new release while the webworkers are still running from the old release, creating the potential for a change to occur post-migration/boostrap, but pre-auditing. I'll link the PR when that is created.

**Comment by gherceg** [comment]
To amend to what I said before, I will still run the bootstrap management command manually on production post-deploy since there is no reason for us to wait the 3 months to create the bootstrap events.

**Pull request merged**